scipy/conftest.py

```diff
@@ -147,10 +148,33 @@ def check_fpu_mode(request):
     reason="do not run with Array API on",
 )
 
-skip_if_array_api_gpu = pytest.mark.skipif(
-    SCIPY_ARRAY_API and SCIPY_DEVICE != 'cpu',
-    reason="do not run with Array API on and not on CPU",
-)
+
+def skip_if_array_api_gpu(func):
+    reason = "do not run with Array API on and not on CPU"
+    # method gets there as a function so we cannot use inspect.ismethod
+    if '.' in func.__qualname__:
+        @wraps(func)
+        def wrapped(self, *args, **kwargs):
+            xp = kwargs["xp"]
+            if SCIPY_ARRAY_API and SCIPY_DEVICE != 'cpu':
+                if xp.__name__ == 'cupy':
+                    pytest.skip(reason=reason)
+                elif xp.__name__ == 'torch':
+                    if torch.cuda.is_available():
+                        pytest.skip(reason=reason)
+            return func(self, *args, **kwargs)
+    else:
+        @wraps(func)
+        def wrapped(*args, **kwargs):
+            xp = kwargs["xp"]
+            if SCIPY_ARRAY_API and SCIPY_DEVICE != 'cpu':
+                if xp.__name__ == 'cupy':
+                    pytest.skip(reason=reason)
+                elif xp.__name__ == 'torch':
+                    if torch.cuda.is_available():
+                        pytest.skip(reason=reason)
+            return func(*args, **kwargs)
+    return wrapped
```

Comment:
`python dev.py test -j 32 -b all -s cluster` does seem a bit more sensible here vs. `main` with 50 tests skipped on this branch vs. 389 on `main`.

Reply:
Strangely, while I get 389 skips on `main`, I only get 32 skips here (with 110 passes). Could another dev run this so we can see which number is the odd one out?